File: src/main/python/g0001_0100/s0045_jump_game_ii/Solution.py
# #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Greedy
# #Algorithm_II_Day_13_Dynamic_Programming #Dynamic_Programming_I_Day_4
# #Big_O_Time_O(n)_Space_O(1) #2024_06_08_Time_89_ms_(99.30%)_Space_17.7_MB_(41.04%)

class Solution:
    def jump(self, nums: List[int]) -> int:
        length = 0
        max_length = 0
        min_jump = 0
        
        for i in range(len(nums) - 1):
            length -= 1
            max_length -= 1
            max_length = max(max_length, nums[i])
            if length <= 0:
                length = max_length
                min_jump += 1
            if length >= len(nums) - i - 1:
                return min_jump
        
        return min_jump

File: src/main/dart/g0001_0100/s0045_jump_game_ii/Solution.dart
// #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Greedy
// #Algorithm_II_Day_13_Dynamic_Programming #Dynamic_Programming_I_Day_4
// #Top_Interview_150_Array/String #Big_O_Time_O(n)_Space_O(1)
// #2024_10_04_Time_335_ms_(81.58%)_Space_148.6_MB_(76.32%)

import 'dart:math';

class Solution {
  int jump(List<int> nums) {
    int length = 0;
    int maxLength = 0;
    int minJump = 0;

    for (int i = 0; i < nums.length - 1; ++i) {
      length--;
      maxLength--;
      maxLength = max(maxLength, nums[i]);

      if (length <= 0) {
        length = maxLength;
        minJump++;
      }

      if (length >= nums.length - i - 1) {
        return minJump;
      }
    }

    return minJump;
  }
}
